### PR TITLE
XTDB-1889 S3 transfer manager

### PR DESCRIPTION
**Pull request by jsulmont**
* added support for S3 transfer manager.

### CODE REVIEW

File: docs/administration/modules/ROOT/pages/checkpointing.adoc [hunk 1]
```diff
@@ -84,3 +84,17 @@ EDN::
 
 == `FileSystem` Checkpoint Store parameters
 * `path` (required, string/`File`/`Path`/`URI`): path to store checkpoints.
+
+=== Note about using `FileSystem` Checkpoint Store
+
+Most Cloud providers offer some high-performance block storage facility (e.g.
```
Comment (jarohen) [inline docs/administration/modules/ROOT/pages/checkpointing.adoc]: [adoc convention](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line) is one sentence per line rather than wrapping - I tend to then use `visual-line-mode` in emacs to soft-wrap them again.

File: modules/s3/src/xtdb/s3/checkpoint.clj [hunk 2]
```diff
@@ -86,10 +149,14 @@
                                        :doc "S3 bucket"}
                               :prefix {:required? false,
                                        :spec ::s3/prefix
-                                       :doc "S3 prefix"}}}
-  [{:keys [^S3Configurator configurator bucket prefix]}]
+                                       :doc "S3 prefix"}
+                              :transfer-manager? {:required? false,
```
Comment (jarohen) [inline modules/s3/src/xtdb/s3/checkpoint.clj]: one to discuss: is it worth us having two modules for this, rather than switching on this flag in a number of places?

e.g. so that a user would configure it as
```clojure
:checkpointer
{:xtdb/module `xtdb.checkpoint/->checkpointer
 :store {:xtdb/module `xtdb.s3.checkpoint/->xfmgr-cp-store
         :bucket "checkpoint-bucket"
         :prefix "checkpoint-dir"
         :configurator `s3-configurator}
 :approx-frequency (java.time.Duration/ofSeconds 30)}
```

File: modules/rocksdb/src/xtdb/rocksdb.clj [hunk 3]
```diff
@@ -393,3 +393,10 @@
                                 :->column-family-handle ->column-family-handle})]
     (cond-> kv-store
       checkpointer (assoc :cp-job (cp/start checkpointer kv-store {::cp/cp-format cp-format})))))
+
```
Comment (jarohen) [inline modules/rocksdb/src/xtdb/rocksdb.clj]: unrelated?

File: modules/s3/src/xtdb/s3.clj [hunk 4]
```diff
@@ -57,7 +59,7 @@
                           (when-let [resp (.get fut)]
                             [path resp]))))))
 
-(defn ^:no-doc list-objects [{:keys [^S3Configurator configurator ^S3AsyncClient client bucket prefix]}
+(defn ^:no-doc list-objects [{:keys [^S3Configurator _configurator ^S3AsyncClient client bucket prefix]}
```
Comment (jarohen) [inline modules/s3/src/xtdb/s3.clj]: nit: can just kill it if it's not used

File: docs/storage/modules/ROOT/pages/aws-s3.adoc [hunk 5]
```diff
@@ -121,3 +124,51 @@ Clojure::
...
+link:https://aws.amazon.com/blogs/developer/introducing-crt-based-s3-client-and-the-s3-transfer-manager-in-the-aws-sdk-for-java-2-x/[CRT-based S3 client] in order to gain its full benefit.
+The following snippet
+
+[tabs]
+====
+Clojure::
++
+[source,clojure]
+----
+(defn- s3-configurator [_]
```
Comment (jarohen) [inline docs/storage/modules/ROOT/pages/aws-s3.adoc:179]: can we give users a utility for this? maybe (see below comment) if we make a different module for the xfer-mgr we can give it a different default for S3 configurator too

**Review by jarohen — COMMENTED** [review]
Cheers @jsulmont, looking good :pray: some minor feedback below, let's catch up

File: docs/storage/modules/ROOT/pages/aws-s3.adoc (hunk 5 above)
Comment (jsulmont) [reply docs/storage/modules/ROOT/pages/aws-s3.adoc:179]: Ok --  do you reckon we should use the CRT S3 Async client by default? That would be the only sensible configuration.  But we would  then have to embark the CRT jar in project.clj (module/s3/project.clj).
Not sure it's the right thing to do. The doc gives an example of how to use `S3Configurator` to configure a CRT Client with options.
I'd think someone having to deal with this would refer to the AWS CRT S3 Async Client documentation. No?
